content_scripts/mode_visual_edit.coffee

```diff
@@ -0,0 +1,789 @@
+      if @initialRangeOnLaunch? and event?.type == "keydown" and KeyboardUtils.isEscape event
+        if @initialRangeOnLaunch == "None"
+          @collapseSelectionToAnchor()
+        else
+          @setSelectionRange @initialRangeOnLaunch unless @initialRangeOnLaunch == "None"
+
+  # Yank the selection; always exits; either deletes the selection or collapses it; set @yankedText and
+  # returns it.
+  yank: (args = {}) ->
+    @yankedText = @selection.toString()
+
+    if @options.deleteFromDocument or args.deleteFromDocument
+      @selection.deleteFromDocument()
+    else
+      @collapseSelectionToAnchor()
+
+    message = @yankedText.replace /\s+/g, " "
+    message = message[...12] + "..." if 15 < @yankedText.length
+    plural = if @yankedText.length == 1 then "" else "s"
+    HUD.showForDuration "Yanked #{@yankedText.length} character#{plural}: \"#{message}\".", 2500
+
+    @options.onYank.call @, @yankedText if @options.onYank
+    @exit()
+    @yankedText
+
+  # For "daw", "das", and so on.  We select a lexical entity (a word, a sentence or a paragraph).
+  # Note(smblott).  It would be better if the entities could be handled symmetrically.  Unfortunately, they
+  # cannot, and we have to handle each case individually.
+  selectLexicalEntity: (entity, count = 1) ->
+
+    switch entity
+      when word
+        if @nextCharacterIsWordCharacter()
+          @runMovements [ forward, character ], [ backward, word ]
+          @collapseSelectionToFocus()
+        @runMovements ([0...count].map -> [ forward, vimword ])...
+
+      when sentence
+        @runMovements [ forward, character ], [ backward, sentence ]
+        @collapseSelectionToFocus()
+        @runMovements ([0...count].map -> [ forward, sentence ])...
+
+      when paragraph
+        # Chrome's paragraph movements are weird: they're not symmetrical, and tend to stop in odd places
+        # (like mid-paragraph, for example).  Here, we define a paragraph as a new-line delimited entity,
+        # including the terminating newline.
+        # Note(smblott).  This does not currently use the count.
+        char = @getNextBackwardCharacter()
+        while char? and char != "\n"
+          return unless @runMovements [ backward, character ], [ backward, lineboundary ]
+          char = @getNextBackwardCharacter()
+        @collapseSelectionToFocus()
+        char = @getNextForwardCharacter()
+        while char? and char != "\n"
+          return unless @runMovements [ forward, character ], [ forward, lineboundary ]
+          char = @getNextForwardCharacter()
+        @runMovement forward, character
+
+  # Try to scroll the focus into view.
+  scrollIntoView: ->
+    @protectClipboard =>
+      if @element and DomUtils.isEditable @element
+        if @element.clientHeight < @element.scrollHeight
+          if @element.isContentEditable
+            # Help(smblott)! How do we do this?  This case matters for Gmail and Google's inbox.
```

Comment:
[`window.getSelection().getRangeAt(0).getClientRects()[0]`](https://developer.mozilla.org/en-US/docs/Web/API/Range.getClientRects) looks like it does the right thing for this, at least on Chrome for android.

The non-editable element case below should probably also use this somewhere so we scroll to the selection in `overflow: scroll` elements.